### PR TITLE
Always create certmanager/cluster CR

### PR DESCRIPTION
**Pull request by slaskawi**
https://issues.redhat.com/browse/AUTH-105

This PR enables the Operator to always create the `certmanager/cluster` CR, so that other pieces of code could assume it's always there. This approach has stronger guarantees than defining the `certmanager/cluster` in the OLM bundle (someone can just delete it). The convention of having those objects comes down from library-go and core. Most of the controllers need it to write their status. 

This decision has a few consequences:
- We no longer need an example in the bundle. 
- We don't need to mention this requirement anywhere in the docs (but I'm doing it either way - to make users aware that there is something like this).

### CODE REVIEW

File: pkg/operator/operatorclient/operatorclient.go [hunk 1]
```diff
@@ -34,12 +36,17 @@ func (c OperatorClient) Informer() cache.SharedIndexInformer {
 	return c.Informers.Operator().V1alpha1().CertManagers().Informer()
 }
 
+// GetOperatorState returns the Operator state. If it's missing (e.g. the certmanager CR hasn't been creates),
```
Comment (s-urbaniak) [inline pkg/operator/operatorclient/operatorclient.go:39]: nit: let's leave out the brackets

File: pkg/operator/operatorclient/operatorclient.go [hunk 2]
```diff
@@ -34,12 +36,17 @@ func (c OperatorClient) Informer() cache.SharedIndexInformer {
...
+// it will create a default one.
 func (c OperatorClient) GetOperatorState() (*operatorv1.OperatorSpec, *operatorv1.OperatorStatus, string, error) {
 	instance, err := c.Informers.Operator().V1alpha1().CertManagers().Lister().Get("cluster")
+	if apierrors.IsNotFound(err) {
+		//TODO: Consider using propagated context here but this requires changes in library-go. Maybe it's not worth it?
+		instance, err = c.createDefaultCertManager(context.TODO())
```
Comment (s-urbaniak) [inline pkg/operator/operatorclient/operatorclient.go:45]: As discussed I believe we have the open question how this can be removed at uninstall time.

File: pkg/operator/operatorclient/operatorclient.go [hunk 3]
```diff
@@ -34,12 +36,17 @@ func (c OperatorClient) Informer() cache.SharedIndexInformer {
 	return c.Informers.Operator().V1alpha1().CertManagers().Informer()
 }
 
+// GetOperatorState returns the Operator state. If it's missing (e.g. the certmanager CR hasn't been creates),
+// it will create a default one.
 func (c OperatorClient) GetOperatorState() (*operatorv1.OperatorSpec, *operatorv1.OperatorStatus, string, error) {
```
Comment (s-urbaniak) [inline pkg/operator/operatorclient/operatorclient.go:41]: It seems strange to me that `GetOperator` state has side effects. Do we have a better place?

File: pkg/operator/operatorclient/operatorclient.go (hunk 2 above)
Comment (slaskawi) [reply pkg/operator/operatorclient/operatorclient.go:45]: If you're uninstalling the Operator? I think it will be removed by OLM, but let me double check this scenario.

File: pkg/operator/operatorclient/operatorclient.go (hunk 3 above)
Comment (slaskawi) [reply pkg/operator/operatorclient/operatorclient.go:41]: Yeah... thinking about this a bit more it's weird. Let me introduce a new Controller to ensure it's always there. It will be cleanest solution I think.

File: pkg/operator/operatorclient/operatorclient.go (hunk 1 above)
Comment (slaskawi) [reply pkg/operator/operatorclient/operatorclient.go:39]: I'm not touching this file anymore.

File: pkg/operator/operatorclient/operatorclient.go (hunk 3 above)
Comment (slaskawi) [reply pkg/operator/operatorclient/operatorclient.go]: Done

File: pkg/controller/deployment/default_cert_manager_creator.go [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+		controllerFactory: factory.New().ResyncEvery(time.Minute).WithInformers(
+			operatorClient.Informer(),
+		),
+		recorder: eventsRecorder.WithComponentSuffix("default-cert-manager-controller"),
+		name:     controllerName,
+	}
+}
+
+func (c *DefaultCertManagerController) Run(ctx context.Context, workers int) {
+	c.controllerFactory.WithSync(c.Sync).ToController("DefaultCertManager_"+c.name, c.recorder).Run(ctx, workers)
```
Comment (s-urbaniak) [inline pkg/controller/deployment/default_cert_manager_creator.go:43]: nit: you want to return the Controller in the `NewDefaultCertManagerController` constructor and remove this method, i.e.
```
	c := &DefaultCertManagerController{
		operatorClient:    operatorClient,
		certManagerClient: certManagerClient,
		controllerFactory: factory.New().ResyncEvery(time.Minute).WithInformers(
			operatorClient.Informer(),
		),
		recorder: eventsRecorder.WithComponentSuffix("default-cert-manager-controller"),
		name:     controllerName,
	}
	return c.controllerFactory.WithSync(c.Sync).ToController("DefaultCertManager_"+c.name, c.recorder)
```

File: pkg/controller/deployment/default_cert_manager_creator.go [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+}
+
+func (c *DefaultCertManagerController) Sync(ctx context.Context, syncCtx factory.SyncContext) error {
```
Comment (s-urbaniak) [inline pkg/controller/deployment/default_cert_manager_creator.go:46]: let's not make this method public. the controller factory `Sync` is.

File: pkg/controller/deployment/default_cert_manager_creator.go (hunk 4 above)
Comment (slaskawi) [reply pkg/controller/deployment/default_cert_manager_creator.go]: Fixed.

File: pkg/controller/deployment/default_cert_manager_creator.go (hunk 5 above)
Comment (slaskawi) [reply pkg/controller/deployment/default_cert_manager_creator.go]: Fixed.

File: pkg/controller/deployment/default_cert_manager_creator.go [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+		operatorClient:    operatorClient,
+		certManagerClient: certManagerClient,
+		controllerFactory: factory.New().ResyncEvery(time.Minute).WithInformers(
+			operatorClient.Informer(),
+		),
+		recorder: eventsRecorder.WithComponentSuffix("default-cert-manager-controller"),
+		name:     controllerName,
+	}
+
+	return controller.controllerFactory.WithSync(controller.sync).ToController("DefaultCertManager_"+controller.name, controller.recorder)
```
Comment (s-urbaniak) [inline pkg/controller/deployment/default_cert_manager_creator.go:41]: let's stay consistent with CamelCase

File: pkg/controller/deployment/default_cert_manager_creator.go [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+
+type DefaultCertManagerController struct {
+	operatorClient    v1helpers.OperatorClient
+	controllerFactory *factory.Factory
+	recorder          events.Recorder
+	name              string
+	certManagerClient alpha1.OperatorV1alpha1Interface
+}
+
+func (c *DefaultCertManagerController) Name() string {
```
Comment (s-urbaniak) [inline pkg/controller/deployment/default_cert_manager_creator.go:26]: curious: why is this needed?

File: pkg/controller/deployment/default_cert_manager_creator.go [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+	return c.name
+}
+
+func NewDefaultCertManagerController(controllerName string, operatorClient v1helpers.OperatorClient, certManagerClient alpha1.OperatorV1alpha1Interface, eventsRecorder events.Recorder) factory.Controller {
```
Comment (s-urbaniak) [inline pkg/controller/deployment/default_cert_manager_creator.go:30]: I'd vote for removing `controllerName`.

File: pkg/controller/deployment/default_cert_manager_creator.go (hunk 6 above)
Comment (s-urbaniak) [reply pkg/controller/deployment/default_cert_manager_creator.go:41]: Actually, I suggest just to leave `DefaultCertManager`.

File: pkg/controller/deployment/default_cert_manager_creator.go [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+	controller := DefaultCertManagerController{
+		operatorClient:    operatorClient,
+		certManagerClient: certManagerClient,
+		controllerFactory: factory.New().ResyncEvery(time.Minute).WithInformers(
+			operatorClient.Informer(),
+		),
+		recorder: eventsRecorder.WithComponentSuffix("default-cert-manager-controller"),
```
Comment (s-urbaniak) [inline pkg/controller/deployment/default_cert_manager_creator.go:37]: I suggest to just leave `default-cert-manager`.

File: pkg/controller/deployment/default_cert_manager_creator.go (hunk 6 above)
Comment (slaskawi) [reply pkg/controller/deployment/default_cert_manager_creator.go:41]: Fixed.

File: pkg/controller/deployment/default_cert_manager_creator.go (hunk 7 above)
Comment (slaskawi) [reply pkg/controller/deployment/default_cert_manager_creator.go:26]: Required by the `factory.Controller` interface.

File: pkg/controller/deployment/default_cert_manager_creator.go (hunk 8 above)
Comment (slaskawi) [reply pkg/controller/deployment/default_cert_manager_creator.go]: Removed.

File: pkg/controller/deployment/default_cert_manager_creator.go (hunk 7 above)
Comment (s-urbaniak) [reply pkg/controller/deployment/default_cert_manager_creator.go]: this is returning an instance created by the controller factory, `Name()` is implemented by `baseController`. The `Name()` function here has no effect.

Comment (s-urbaniak) [reply pkg/controller/deployment/default_cert_manager_creator.go]: https://github.com/openshift/cert-manager-operator/blob/b9570ecc15835f9e3b266957f0e0be01b97e1a09/vendor/github.com/openshift/library-go/pkg/controller/factory/base_controller.go#L45-L47

Comment (slaskawi) [reply pkg/controller/deployment/default_cert_manager_creator.go]: Removed
